packages/user/TokenStream/service/src/tests.rs

```diff
@@ -0,0 +1,413 @@
+    }
+
+    fn setup_env(chain: &psibase::Chain) -> Result<u32, psibase::Error> {
+        chain.new_account(ALICE)?;
+        chain.new_account(BOB)?;
+        chain.new_account(CHARLIE)?;
+
+        let supply = Quantity::from(10000000);
+
+        let token_id = Tokens::push_from(&chain, ALICE)
+            .create(4.try_into().unwrap(), supply)
+            .get()
+            .unwrap();
+        assert_eq!(token_id, 2);
+
+        Tokens::push_from(&chain, ALICE).mint(
+            token_id,
+            supply,
+            "memo".to_string().try_into().unwrap(),
+        );
+        Ok(token_id)
+    }
+
+    pub fn full_vesting_time(half_life_seconds: u32, balance: u64) -> i64 {
+        if balance == 0 {
+            return 0;
+        }
+        let t = half_life_seconds as f64 * ((2.0 * balance as f64).ln() / std::f64::consts::LN_2);
+        t.ceil() as i64 + 1
+    }
+
+    fn reset_clock(chain: &psibase::Chain) {
+        let time: i64 = 1000;
+        chain.start_block_at(TimePointSec { seconds: time }.microseconds());
+    }
+
+    fn create_stream(
+        chain: &psibase::Chain,
+        author: AccountNumber,
+        half_life_seconds: u32,
+        token_id: u32,
+    ) -> u32 {
+        TokenStream::push_from(&chain, author)
+            .create(half_life_seconds, token_id)
+            .get()
+            .unwrap()
+    }
+
+    #[psibase::test_case(packages("TokenStream"))]
+    fn test_basics(mut chain: psibase::Chain) -> Result<(), psibase::Error> {
+        chain.set_auto_block_start(false);
+        chain.start_block();
+
+        reset_clock(&chain);
+
+        let token_id = setup_env(&chain)?;
+
+        tokens_credit(&chain, token_id, ALICE, TOKEN_STREAM, 500);
+
+        let stream_nft_id = create_stream(&chain, ALICE, 10000, token_id);
+
+        check_balance(&chain, token_id, TOKEN_STREAM, Some(0));
+
+        assert_eq!(
+            get_shared_balance(&chain, token_id, ALICE, TOKEN_STREAM),
+            500.into()
+        );
+
+        TokenStream::push_from(&chain, ALICE)
+            .deposit(stream_nft_id)
+            .get()
+            .unwrap();
+
+        assert_eq!(
+            get_stream(&chain, stream_nft_id).total_deposited,
+            500.into()
+        );
+
+        Nfts::push_from(&chain, ALICE).credit(stream_nft_id, BOB, "memo".to_string());
+        Nfts::push_from(&chain, BOB).debit(stream_nft_id, "memo".to_string());
+
+        // Claim fails because no vesting has yet occured.
+        assert_eq!(
+            TokenStream::push_from(&chain, BOB)
+                .claim(stream_nft_id)
+                .get()
+                .unwrap_err()
+                .to_string(),
+            "service 'tokens' aborted with message: credit quantity must be greater than 0"
+                .to_string()
+        );
+
+        check_balance(&chain, token_id, BOB, Some(0));
+
+        let expected_full_vesting_wait_time = full_vesting_time(10000, 500);
+
+        chain.start_block_at(
```

Comment:
Should this be `start_block_after`? The initial time set by `reset_clock` doesn't seem to be taken into account.